### PR TITLE
sub_select fix #2

### LINKED ISSUE #151: RFC: `sub_select` should have a nullable result type by default
tl;dr: `sub_select` should have the type `sub_select :: SqlQuery (SqlExpr (Value a)) -> SqlExpr (Value (Maybe a))`.

We use `sub_select` extensively in the work codebase, and it is a common source of runtime errors. The general pattern is this:

```haskell
bad :: SqlPersistM [(Entity Foo, XPropType)]
bad = 
  select $
  from $ \foo -> do 
  where_ (foo ^. FooProperty ==. val 3)
  pure (foo, sub_select (from_ \x -> where_ (p x) >> limit 1 >> pure (x ^. XProp))
```

`sub_select` has the type `sub_select :: SqlQuery (SqlExpr (Value a)) -> SqlExpr (Value a)`. Because the result type is `Value a`, the relevant marshalling code assumes that it cannot be `NULL`. However, it totally can be `NULL`, for a huge amount of the potential uses - possibly even the majority or default case.

A problem is that it is going to be quite difficult to statically determine whether or not it will return null. `sub_select $ from $ \x -> where_ (p x) >> limit 1 >> pure (x ^. XProp)` will obviously return `NULL` if the query returns 0 results, and `persistent` will fail to parse a `Nothing`. The work-around is to use `just`.

However, `sub_select $ from $ \x -> where_ (p x) >> pure (sum (x ^. XProp))` will return `0` for zero rows, so it doesn't *need* to be `Maybe`. Likewise, `sub_select $ from $ \x -> where_ (p x) >> pure countRows` will also return a non-null result.

I propose that `sub_select` be made nullable by default, along with another function that can dismiss the `Maybe` when the user knows exactly what they're doing and it'll be fine.

### PR DESCRIPTION
**Pull request by parsonsmatt**
Fixes #151 

This PR adds a deprecation notice and suggests a migration path away.

image

### CODE REVIEW

File: src/Database/Esqueleto/Internal/Internal.hs [hunk 1]
```diff
@@ -339,12 +339,79 @@ having expr = Q $ W.tell mempty { sdHavingClause = Where expr }
...
+-- 1. If the query passed has 0 result rows, then it will return a @NULL@ value.
+--    The @persistent@ parsing operations will fail on an unexpected @NULL@.
+-- 2. If the query passed returns more than one row, then the SQL engine will
+--    fail with an error like "More than one row returned by a subquery used as
+--    an expression".
+--
+-- This function is safe if you guarantee that exactly one row will be returned,
+-- or if the result already has a 'Maybe' type for some reason.
+--
+-- For variatns with the safety encoded already, see 'subSelect' and
```
[the remark this answers is not in the document]
Comment (parsonsmatt) [reply src/Database/Esqueleto/Internal/Internal.hs]: Thanks for catching that!

**Comment by Vlix** [comment]
What's the reasoning behind the default `LIMIT 1`?
We mostly have used `sub_select` in `IN` queries. (i.e. `WHERE x IN (SELECT field FROM table)`) and it seems very very wrong to have it limited to 1 result. And having this use case use `subSelectUnsafe` seems weird, since this is a very normal use of a sub select, no?

I dunno, it just doesn't feel right. If you add the `LIMIT 1`, that sounds like it's a sub select used in a specific case, is all I'm saying. Might want to have `subSelect` and `subSelectOne` or something?

**Comment by parsonsmatt** [comment]
Thanks for the feedback, that's very helpful! We don't have any uses following that pattern in our codebase.

`in_` is typed as 

```haskell
in_ :: PersistField typ => SqlExpr (Value typ) -> SqlExpr (ValueList typ) -> SqlExpr (Value Bool)
```

So it seems like it requires `subList_select`:

```haskell
-- | Execute a subquery @SELECT@ in an SqlExpression.  Returns a
-- list of values.
subList_select :: PersistField a => SqlQuery (SqlExpr (Value a)) -> SqlExpr (ValueList a)
subList_select         = EList . sub_select
```

This types it as returning multiple potential results. So I'm not sure that including `LIMIT 1` in the `subSelect` function is unsafe, though I do think a `subSelectList` to maintain symmetry with the other `subSelectX` functions would be good.

**Comment by Vlix** [comment]
Ah, you are right, we've used `subList_select` exclusively with `notIn` or `in_`.
So as long as there's a `subSelectList` or anything similar, I guess this should be fine.

I'm still wondering if we could get a sub select in the form of `SELECT * FROM (SELECT * FROM table ...) a WHERE a.field = val` to work in some way with `esqueleto`, but that would probably be a different kind of `from` instead of a kind of `sub_select`, right?

**Comment by parsonsmatt** [comment]
I haven't figured out how to add new things to the FROM clause. That would be very nice to have.

File: test/Common/Test.hs [hunk 2]
```diff
@@ -1,5 +1,5 @@
 {-# OPTIONS_GHC -fno-warn-unused-binds  #-}
-{-# OPTIONS_GHC -fno-warn-deprecations  #-}
+-- {-# OPTIONS_GHC -fno-warn-deprecations  #-}
```
Comment (bitemyapp) [inline test/Common/Test.hs]: Planning to flip it back on?

If not, just delete IMO.

**Comment by parsonsmatt** [comment]
Tests added @bitemyapp looks good to go?

**Pull request merged**